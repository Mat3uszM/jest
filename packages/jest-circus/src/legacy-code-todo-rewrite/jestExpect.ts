/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import expect = require('expect');

import {
  addSerializer,
  toMatchInlineSnapshot,
  toMatchSnapshot,
  toThrowErrorMatchingInlineSnapshot,
  toThrowErrorMatchingSnapshot,
} from 'jest-snapshot';

export default (config: {
  expand: boolean;
  ignoreAsymmetricMatches: boolean;
}) => {
  global.expect = expect;
  expect.setState({
    expand: config.expand,
    ignoreAsymmetricMatches: config.ignoreAsymmetricMatches,
  });
  expect.extend({
    toMatchInlineSnapshot,
    toMatchSnapshot,
    toThrowErrorMatchingInlineSnapshot,
    toThrowErrorMatchingSnapshot,
  });

  expect.addSnapshotSerializer = addSerializer;
};
