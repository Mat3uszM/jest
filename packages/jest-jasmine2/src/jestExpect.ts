/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

/* eslint-disable local/prefer-spread-eventually */

import type {Expect, Global} from '@jest/types';
import expect = require('expect');
import {
  addSerializer,
  toMatchInlineSnapshot,
  toMatchSnapshot,
  toThrowErrorMatchingInlineSnapshot,
  toThrowErrorMatchingSnapshot,
} from 'jest-snapshot';
import type {Jasmine, JasmineMatchersObject} from './types';

declare const global: Global.Global;

export default (config: {expand: boolean}): void => {
  const jestExpect = expect as Expect.JestExpect;

  global.expect = jestExpect;
  jestExpect.setState({expand: config.expand});
  jestExpect.extend({
    toMatchInlineSnapshot,
    toMatchSnapshot,
    toThrowErrorMatchingInlineSnapshot,
    toThrowErrorMatchingSnapshot,
  });
  jestExpect.addSnapshotSerializer = addSerializer;

  const jasmine = global.jasmine as Jasmine;
  jasmine.anything = jestExpect.anything;
  jasmine.any = jestExpect.any;
  jasmine.objectContaining = jestExpect.objectContaining;
  jasmine.arrayContaining = jestExpect.arrayContaining;
  jasmine.stringMatching = jestExpect.stringMatching;

  jasmine.addMatchers = (jasmineMatchersObject: JasmineMatchersObject) => {
    const jestMatchersObject = Object.create(null);
    Object.keys(jasmineMatchersObject).forEach(name => {
      jestMatchersObject[name] = function (
        this: Expect.MatcherState,
        ...args: Array<unknown>
      ): Expect.RawMatcherFn {
        // use "expect.extend" if you need to use equality testers (via this.equal)
        const result = jasmineMatchersObject[name](null, null);
        // if there is no 'negativeCompare', both should be handled by `compare`
        const negativeCompare = result.negativeCompare || result.compare;

        return this.isNot
          ? negativeCompare.apply(
              null,
              // @ts-expect-error
              args,
            )
          : result.compare.apply(
              null,
              // @ts-expect-error
              args,
            );
      };
    });

    jestExpect.extend(jestMatchersObject);
  };
};
