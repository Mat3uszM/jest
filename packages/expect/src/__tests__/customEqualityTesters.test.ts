/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import type {Tester} from '@jest/expect-utils';
import jestExpect from '../';

const specialObjPropName = '$$special';
const specialObjSymbol = Symbol('special test object type');

interface SpecialObject {
  $$special: symbol;
  value: number;
}

function createSpecialObject(value: number) {
  return {
    [specialObjPropName]: specialObjSymbol,
    value,
  };
}

function isSpecialObject(a: unknown): a is SpecialObject {
  return (
    a != null &&
    typeof a === 'object' &&
    specialObjPropName in a &&
    (a as any)[specialObjPropName] === specialObjSymbol
  );
}

const specialObjTester: Tester = (
  a: unknown,
  b: unknown,
): boolean | undefined => {
  const isASpecial = isSpecialObject(a);
  const isBSpecial = isSpecialObject(b);

  if (isASpecial && isBSpecial) {
    return true;
  } else if ((isASpecial && !isBSpecial) || (!isASpecial && isBSpecial)) {
    return false;
  } else {
    return undefined;
  }
};

function* toIterator<T>(array: Array<T>): Iterator<T> {
  for (const obj of array) {
    yield obj;
  }
}

it('has no custom testers as default', () => {
  const special1 = createSpecialObject(1);
  const special2 = createSpecialObject(2);

  // Basic matchers passing
  expect(special1).toBe(special1);
  expect(special1).toEqual(special1);
  expect([special1, special2]).toEqual([special1, special2]);
  expect(new Set([special1])).toEqual(new Set([special1]));
  expect(new Map([['key', special1]])).toEqual(new Map([['key', special1]]));
  expect(toIterator([special1, special2])).toEqual(
    toIterator([special1, special2]),
  );

  // Basic matchers not passing
  expect(special1).not.toBe(special2);
  expect(createSpecialObject(1)).not.toEqual(createSpecialObject(2));
  expect([special1, special2]).not.toEqual([special2, special1]);
  expect(new Set([special1])).not.toEqual(new Set([special2]));
  expect(new Map([['key', special1]])).not.toEqual(
    new Map([['key', special2]]),
  );
  expect(toIterator([special1, special2])).not.toEqual(
    toIterator([special2, special1]),
  );
});

describe('with custom equality testers', () => {
  let originalTesters: Array<Tester>;

  beforeAll(() => {
    originalTesters = jestExpect.customEqualityTesters;
    jestExpect.customEqualityTesters = [...originalTesters, specialObjTester];
  });

  afterAll(() => {
    jestExpect.customEqualityTesters = originalTesters;
  });

  it('basic matchers customTesters do not apply to', () => {
    const special1 = createSpecialObject(1);
    const special2 = createSpecialObject(2);

    expect(special1).toBe(special1);
    expect([special1]).toContain(special1);

    expect(special1).not.toBe(special2);
    expect([special1]).not.toContain(special2);
  });

  it('basic matchers customTesters do apply to', () => {
    const special1 = createSpecialObject(1);
    const special2 = createSpecialObject(2);

    expect(special1).toEqual(special1);
    expect(special1).toEqual(special2);
    expect([special1, special2]).toEqual([special2, special1]);
    expect(new Map([['key', special1]])).toEqual(new Map([['key', special2]]));
    expect(new Set([special1])).toEqual(new Set([special2]));
    expect(toIterator([special1, special2])).toEqual(
      toIterator([special2, special1]),
    );
    expect([createSpecialObject(1)]).toContainEqual(createSpecialObject(2));
    expect({a: createSpecialObject(1)}).toHaveProperty(
      'a',
      createSpecialObject(2),
    );
  });

  // it('applies custom testers to toStrictEqual', () => {});

  // it('applies custom testers to toMatchObject', () => {});

  // TODO: Add tests for other matchers
  // TODO: Add tests for built-in custom testers (e.g. iterableEquality, subsetObjectEquality)
  // TODO: Add tests for extended expect matchers that use this.equal();
});
