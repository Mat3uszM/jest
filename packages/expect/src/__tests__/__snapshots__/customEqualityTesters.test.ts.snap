// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with custom equality testers toMatchObject error shows special objects as equal 1`] = `
"<dim>expect(</intensity><red>received</color><dim>).</intensity>toMatchObject<dim>(</intensity><green>expected</color><dim>)</intensity>

<green>- Expected  - 1</color>
<red>+ Received  + 1</color>

<dim>  Object {</intensity>
<green>-   "a": 2,</color>
<red>+   "a": 1,</color>
<dim>    "b": Object {</intensity>
<dim>      "$$special": Symbol(special test object type),</intensity>
<dim>      "value": "2",</intensity>
<dim>    },</intensity>
<dim>  }</intensity>"
`;
